### ISSUE TITLE
A new Overlay subpart module

### ISSUE THREAD

**Issue by tech4GT** [publiclab/image-sequencer#1080]
Our current overlay module works perfectly for complete Images but in some cases we just want part of the Image to be overlaid. Like over at image-sequencer-app we are trying to overlay one warped image onto another but that does not work correctly since the area around the warped image is black.
I think we can create a subpart-overlay module which can take the coordinates of the object in the image to overlay and can overlay just that.
The problem this boils down to is making a function which will determine whether a given point is inside the quadrilateral defined by the 4 input points.
I would love to hear your thoughts in this @publiclab/is-reviewers 
cc @jywarren

**Comment by tech4GT** [comment]
https://user-images.githubusercontent.com/25617855/58384650-d4802000-8001-11e9-9497-397ffc203a7d.jpeg

**Comment by jywarren** [comment]
Hi! Is this resolved in https://github.com/publiclab/image-sequencer/pull/1082 ?

**Comment by tech4GT** [comment]
Actually I made it work for that specific use case using a custom overlay module but this is still something we might wanna have. It may come in handy, Don't you think?

**Comment by jywarren** [comment]
Sure can you just write out the additional fix you're proposing? Thanks!

**Comment by tech4GT** [comment]
Sure! I am thinking of using morphological operations to determine the boundary of the object in the image and then only overlay the pixels inside the boundary! Maybe someone over at the soc team can pick this up, the challenge here is that we cannot conventional methods since those with increase the runtime by a factor of image width!! I think exploring the morphological operations route is well worth it! :smile:

**Comment by jywarren** [comment]
> using morphological operations to determine the boundary of the object in the image and then only overlay the pixels inside the boundary

This would be an optimization, right? Are we all the way there with correct final bounding box in the un-optimized mode, first? But yes, once we have the basic version working, an optimization would be great!

**Comment by blurry-x-face** [comment]
Hi @tech4GT could you explain this a bit more, I would like to work on this!

**Comment by tech4GT** [comment]
Hi @blurry-x-face 
So it's been a while but as far as I remember this module was supposed to overlay a part of the image onto another. For instance, say I give the module the four corner points of a quadrilateral, I then want only the image inside this quadrilateral to be overlaid. The challenge, as I explained earlier is to determine which pixel is inside the quadrilateral and which one is not. I did try to come up with a solution but that was infeasible due to the complexity being O(height*width*width) which would take forever to complete.

**Comment by blurry-x-face** [comment]
@tech4GT how about if we crop and then overlay, wouldn't that work?

**Comment by tech4GT** [comment]
Yeah that's the challenging part, how do we crop an arbitrary quadrilateral out of the image using just the four corner points.

**Comment by blurry-x-face** [comment]
I'll try and make something and let you know if anything works out, thanks!

**Comment by blurry-x-face** [comment]
@tech4GT I was able to make this, this runs in O(n^2) and also uses web assembly so its fast!

![Screenshot from 2020-02-01 16-30-18](https://user-images.githubusercontent.com/42492389/73591086-2faece80-4510-11ea-9fa7-1b50aa7f8fec.png)

If this looks good I can open a PR for it!

**Comment by tech4GT** [comment]
Sure! Great work!

**Comment by blurry-x-face** [comment]
Thanks! :smile:
